### PR TITLE
Anondecks UI

### PR DESCRIPTION
**Pull request by DrSharky**

### CODE REVIEW

**Comment by pikulet** [comment]
**give a guideline on picture size i.e. 40x40**

**do not make it compulsory to have an image**

**error message for undefined image does not give index**

image

**should reject save if death message is invalid**

image

**default deck name appears as append**

image

**unable to quote in game** black screen/ react error

image

File: Games/core/Message.js [hunk 1]
```diff
@@ -89,12 +90,110 @@ module.exports = class Message {
...
+
+  adjustColor(color) {
+      return  {
+        darkTheme: this.getIncreasedBrightness(color, "#181a1b"),
+        lightTheme: this.getDecreasedBrightness(color, "#ffffff")
+      };
+  }
+
+  increaseBrightness(color, percent) {
+    let num = parseInt(color.replace("#", ""), 16),
```
Comment (pikulet) [inline Games/core/Message.js:160]: please move 2 utils

File: Games/core/Game.js [hunk 2]
```diff
@@ -1549,19 +1549,24 @@ module.exports = class Game {
         let userId = player.userId || player.user.id;
         let user = await models.User.findOne({ id: userId }).select("_id");
 
-        if (user) users.push(user._id);
+        if (user) {
+          users.push(user._id);
+        }
+        else {
+          users.push(null);
```
Comment (pikulet) [inline Games/core/Game.js:1556]: when joining a game, why would user id be null? Is this for bot accounts

File: Games/core/Message.js [hunk 3]
```diff
@@ -89,7 +91,18 @@ module.exports = class Message {
       (player.alive || (!player.alive && version.sender.alive))
     )
       senderId = "anonymous";
-    else if (version.sender) senderId = version.sender.id;
+    else if (version.sender) {
+      senderId = version.sender.id;
+
+      if (version.sender.anonId !== undefined) {
+        version.textColor = version.sender.user.textColor !== undefined ? utils.adjustColor(version.sender.user.textColor) : "";
```
Comment (pikulet) [inline Games/core/Message.js:98]: why does this use utils.adjustColor, but the else clause use this.adjustColor

File: Games/core/Player.js [hunk 4]
```diff
@@ -517,10 +526,90 @@ module.exports = class Player {
     return quote;
   }
 
+  getIncreasedBrightness(color1, color2) {
```
Comment (pikulet) [inline Games/core/Player.js:529]: move all this to utils

File: Games/core/Player.js [hunk 5]
```diff
@@ -57,10 +58,13 @@ module.exports = class Player {
     };
 
     this.id = shortid.generate();
+    this.anonId = deckProfile.id;
     this.user.id = shortid.generate();
     this.name = deckProfile.name;
-    this.user.avatar = false;
-    delete this.user.textColor;
+    this.user.avatar = true;
```
Comment (pikulet) [inline Games/core/Player.js:64]: need not set to true anymore, the line below does it

File: Games/core/Game.js [hunk 6]
```diff
@@ -1549,19 +1549,24 @@ module.exports = class Game {
...
+      var dbPlayers = players.map((p) => p.id);
 
       var game = new models.Game({
         id: this.id,
         type: this.type,
         lobby: this.lobby,
         setup: setup._id,
         users: users,
-        players: players,
+        players: dbPlayers,
```
Comment (pikulet) [inline Games/core/Game.js:1569]: maybe playerIds is a better name

File: react_main/src/components/Deck.jsx [hunk 7]
```diff
@@ -19,20 +18,16 @@ export default function AnonymousDeck(props) {
...
-  let profiles = JSON.parse(props.deck.profiles).map((p) => (
-    <DeckProfile profile={p} />
-  ));
+  var profiles;
 
   let displayName = `${props.deck.name} (${props.deck.id})`;
   return (
     <div className="deck" ref={deckRef} onClick={onClick}>
       <div className="deck-name">{displayName}</div>
-      {disablePopover && profiles}
```
Comment (pikulet) [inline react_main/src/components/Deck.jsx:35]: we do not need disable popover

File: react_main/src/components/Popover.jsx [hunk 8]
```diff
@@ -561,8 +561,14 @@ export function parseDeckPopover(deck) {
   }
 
   //Words
-  let words = tempParseProfilesToWords(deck.profiles);
-  result.push(<InfoRow title="Words" content={words} key="words" />);
+  // let words = tempParseProfilesToWords(deck.profiles);
```
Comment (pikulet) [inline react_main/src/components/Popover.jsx:564]: can remove old code

File: react_main/src/components/Popover.jsx [hunk 9]
```diff
@@ -561,8 +561,14 @@ export function parseDeckPopover(deck) {
...
+  result.push(<InfoRow title="Profiles" key="profiles" />);
+  deck.profiles.map((profile) => {
+    let avatar = profile.avatar !== undefined;
```
Comment (pikulet) [inline react_main/src/components/Popover.jsx:567]: avatar is a boolean, if profile.avatar is false but not undefined, this result might be wrong?

File: react_main/src/css/deck.css [hunk 10]
```diff
@@ -1,12 +1,16 @@
...
-  align-self: flex-start;
-
+  /* align-self: flex-start; */
   display: flex;
   flex-flow: row nowrap;
-  align-items: center;
+  align-items: flex-start;
+  justify-content: flex-start;
+
+  /* display: flex;
```
Comment (pikulet) [inline react_main/src/css/deck.css:8]: delete unused css

File: react_main/src/pages/Game/Game.jsx [hunk 11]
```diff
@@ -1816,6 +1847,17 @@ export function PlayerRows(props) {
       }
     }
 
+    let avatarId;
+
+  if (player !== undefined) {
```
Comment (pikulet) [inline react_main/src/pages/Game/Game.jsx:1852]: when would player be undefined?

avatarId = player.anonId ?? player.userId

File: react_main/src/pages/Play/Decks/CreateDeck.jsx [hunk 12]
```diff
@@ -1,114 +1,286 @@
...
+        .post("/deck/create", {
+          name: deckName,
+          profiles: profiles,
+          editing: editing,
+          id: params.get("edit"),
+        })
+        .then((res) => {
+          siteInfo.showAlert(
+            `${editing ? "Edited" : "Created"} deck '${deckName}'`, "success");
+          !editing ? setEditing(true) : setEditing(false);
```
Comment (pikulet) [inline react_main/src/pages/Play/Decks/CreateDeck.jsx:94]: editing is definitely false here

File: routes/anonymousDeck.js [hunk 13]
```diff
@@ -52,9 +56,10 @@ router.post("/create", async function (req, res) {
       }
     }
 
-    let deck = Object(req.body);
-    deck.name = String(deck.name || "");
-    deck.profiles = Object(deck.profiles);
+    let deck = Object();
+    deck.editing = Boolean(req.body.editing);
```
Comment (pikulet) [inline routes/anonymousDeck.js:60]: I'm a bit confused - when is the /create req.body.editing used and when is /edit used?

File: routes/anonymousDeck.js [hunk 14]
```diff
@@ -315,12 +645,60 @@ router.get("/yours", async function (req, res) {
...
+    logger.error(e);
+    res.status(500);
+    res.send("Unable to find deck profile.");
+  }
+});
+
+router.get("/profiles/:id", async function (req, res) {
+  res.setHeader("Content-Type", "application/json");
+  try {
+    // Gets all profiles for a deck, given a list of profile ids.
```
Comment (pikulet) [inline routes/anonymousDeck.js:673]: given deck id

File: routes/anonymousDeck.js [hunk 15]
```diff
@@ -315,12 +645,60 @@ router.get("/yours", async function (req, res) {
...
+    let deckId = String(req.params.id);
+    let profiles = await models.AnonymousDeck.findOne({ id: deckId })
```
Comment (pikulet) [inline routes/anonymousDeck.js:675]: this var name should be deck

File: react_main/src/pages/Game/Game.jsx [hunk 16]
```diff
@@ -1343,11 +1371,13 @@ function Message(props) {
           <NameWithAvatar
             dead={playerDead && props.stateViewing > 0}
             id={player.userId}
+            avatarId={avatarId}
             name={player.name}
             avatar={player.avatar}
             color={
-              playerDead && props.stateViewing > 0 ? deadGray : player.nameColor
-            }
+              !user.settings?.ignoreTextColor && message.nameColor !== ""
```
[the remark this answers is not in the document]
Comment (DrSharky) [reply react_main/src/pages/Game/Game.jsx:1367]: I think so? I don't remember the intended purpose of the ignore text color setting, and I don't remember when or who put that in.

File: react_main/src/pages/Game/Game.jsx (hunk 11 above)
Comment (DrSharky) [reply react_main/src/pages/Game/Game.jsx]: Not sure? There's quite a few things left in, probably don't need this.

File: react_main/src/pages/Play/Decks/CreateDeck.jsx [hunk 17]
```diff
@@ -1,114 +1,286 @@
...
+    defaultValues: {
+      defProfile: [{ name: "test", avatar: "", color: "", /*deathMessage: "",*/ id: "", deck: "", preview: "", image: "" }],
     },
-  ]);
+  });
+  const { fields, append, remove } = useFieldArray({
+    name: "cards",
+    control
+  });
+  renderCount++;
```
[the remark this answers is not in the document]
Comment (DrSharky) [reply react_main/src/pages/Play/Decks/CreateDeck.jsx]: which variable, renderCount?

File: react_main/src/pages/Play/Decks/CreateDeck.jsx [hunk 18]
```diff
@@ -1,114 +1,286 @@
...
 
-  function onCreateDeck(editing) {
-    let profiles = tempParseWordsToProfiles(formFields[1].value);
+  const [deckName, setDeckName] = useState("");
+  const [editing, setEditing] = useState(false);
+  const [selectedFiles, setSelectedFiles] = useState([]);
+
+  const setFile = index => e => {
+    console.log(`index: ${index}`);
```
[the remark this answers is not in the document]
Comment (DrSharky) [reply react_main/src/pages/Play/Decks/CreateDeck.jsx:53]: ???

File: react_main/src/pages/Play/Decks/CreateDeck.jsx (hunk 12 above)
Comment (DrSharky) [reply react_main/src/pages/Play/Decks/CreateDeck.jsx:110]: I don't know what you mean.

File: react_main/src/pages/Play/Decks/CreateDeck.jsx [hunk 19]
```diff
@@ -1,114 +1,286 @@
...
+          profiles: profiles,
+          editing: editing,
+          id: params.get("edit"),
+        })
+        .then((res) => {
+          siteInfo.showAlert(
+            `${editing ? "Edited" : "Created"} deck '${deckName}'`, "success");
+          !editing ? setEditing(true) : setEditing(false);
+          history.push({ search: `?edit=${res.data}` });
+          for (let i = 0; i < 5; i++) {
```
[the remark this answers is not in the document]
Comment (DrSharky) [reply react_main/src/pages/Play/Decks/CreateDeck.jsx:112]: I think it has been fixed to add 5 default profiles, if I'm not mistaken?

File: routes/anonymousDeck.js (hunk 13 above)
Comment (DrSharky) [reply routes/anonymousDeck.js:59]: I think there is some useless code that I left in, and I forgot to remove some custom /create profile request functions and I think one for edit. Instead, it's all included in the ones that existed previously now. I think there are some request functions that can be deleted as they aren't used.

**Comment by DrSharky** [comment]
Since there are several things here you've commented on, I'll take another pass and remove stuff I forgot to remove (functions in routes/anonymousDecks.js) and revise the smaller changes you've made in the comments.

**Comment by pikulet** [comment]
Fixes #1136

File: react_main/src/pages/Play/Decks/CreateDeck.jsx (hunk 12 above)
Comment (pikulet) [reply react_main/src/pages/Play/Decks/CreateDeck.jsx:108]: this part of the code is nested under if (!editing) , line 83

File: routes/anonymousDeck.js [hunk 20]
```diff
@@ -108,6 +98,90 @@ router.post("/create", async function (req, res) {
...
+
+    let profiles = await models.AnonymousDeck.findOne({ id: deckId }).select(
+      "profiles"
+    );
+
+    profiles = await models.DeckProfile.find({ _id: { $in: profiles } }).select(
+      "_id name avatar color deathMessage"
+    );
+
+    //Case 1: No profiles exist, so just create new ones.
```
Comment (pikulet) [inline routes/anonymousDeck.js:142]: will this not be created in /profiles/create?

this /edit function looks like it's not doing anything?

and i only see a case#1 here

File: routes/anonymousDeck.js [hunk 21]
```diff
@@ -315,12 +556,62 @@ router.get("/yours", async function (req, res) {
   }
 });
 
+router.get("/profile", async function (req, res) {
```
Comment (pikulet) [inline routes/anonymousDeck.js:559]: is this used anywhere now

File: routes/anonymousDeck.js [hunk 22]
```diff
@@ -108,6 +98,90 @@ router.post("/create", async function (req, res) {
   }
 });
 
+router.post("/edit", async function (req, res) {
```
Comment (pikulet) [inline routes/anonymousDeck.js:101]: this route seems to not be doing anything?

File: react_main/src/pages/User/User.jsx [hunk 23]
```diff
@@ -188,7 +190,13 @@ export function Avatar(props) {
   else if (large) size = "large";
   else size = "";
 
-  if (hasImage && !imageUrl && id) {
+  if (hasImage && !imageUrl && id && avatarId) {
+    if (id === avatarId && !deckProfile) {
+      style.backgroundImage = `url(/uploads/${id}_avatar.jpg?t=${siteInfo.cacheVal})`;
```
Comment (pikulet) [inline react_main/src/pages/User/User.jsx:195]: image
 still seeing this wonky path

File: routes/anonymousDeck.js (hunk 22 above)
Comment (DrSharky) [reply routes/anonymousDeck.js:101]: This should be for editing the deck. You add/remove profiles, then press the checkmark icon, and it posts to this route.

File: routes/anonymousDeck.js (hunk 21 above)
Comment (DrSharky) [reply routes/anonymousDeck.js:561]: I don't think so.

File: react_main/src/pages/User/User.jsx (hunk 23 above)
Comment (pikulet) [reply react_main/src/pages/User/User.jsx:195]: It's the default view when I click create deck

File: routes/anonymousDeck.js [hunk 24]
```diff
@@ -315,13 +534,51 @@ router.get("/yours", async function (req, res) {
   }
 });
 
+router.get("/profiles/:id", async function (req, res) {
```
Comment (pikulet) [inline routes/anonymousDeck.js:537]: this route is not needed i think, same as get deck

File: react_main/src/pages/User/User.jsx (hunk 23 above)
Comment (pikulet) [reply react_main/src/pages/User/User.jsx:195]: image
image

flows to case c but is undefined

Comment (pikulet) [reply react_main/src/pages/User/User.jsx:195]: fixed, default avatar empty string

File: react_main/src/pages/Play/Decks/CreateDeck.jsx [hunk 25]
```diff
@@ -1,114 +1,324 @@
...
+                          {/* <label>
+                    <span>Death Message</span>
+                    <input
+                      type="text"
+                      {...register(`cards.${index}.deathMessage`)}
+                      defaultValue={profile.deathMessage}
+                    />
+                  </label> */}
+                          <label>
+                            <span>Delete?</span>
```
Comment (pikulet) [inline react_main/src/pages/Play/Decks/CreateDeck.jsx:208]: deleting profiles resets state to last save of deck
[](file:///home/joyce/Videos/Screencasts/Screencast%20from%2007-09-2023%2023:25:15.webm)

File: routes/game.js [hunk 26]
```diff
@@ -449,15 +449,31 @@ router.post("/host", async function (req, res) {
...
+        function (err, profiles) {
+          if (err) {
+            res.status(500);
+            res.send("Unable to find profiles.");
+            return;
+          }
+
+          if (deck.profiles.length < setup.total) {
+            res.status(500);
+            res.send("This deck is too small for the chosen setup.");
```
Comment (pikulet) [inline routes/game.js:463]: You only exit the inner function, so the game continues to get hosted

**Pull request merged**